Bug 654

rzezeski opened the pull request:
A proposed patch for bug-654.

I created the generic function extract_headers so that adding other custom headers down the line (for example, to set the tree order) would be simple.  E.g. just add {tree_order, ?HEAD_TREE_ORDER} to the list and it's good to go.

I have no logical reason for choosing a custom header vs. a query param...it just _felt_ better to me.  I'd be happy to add or change this to use a query param instead.

rzezeski committed:
allow block size to be set on creation

Allow the block_size property of a file to be set on creation via the
HTTP resource by way of the X-Luwak-Block-Size header.

beerriot commented:
The header value needs a little more frobbing: luwak_file expects block_size to be an integer, but this code sets it as a string.

rzezeski committed:
integer expected for block size -- add a cast

rzezeski commented:
On my first patch I simply did a curl and checked the file object, but didn't verify the contents.  This time I was a little more thorough; I went so far as to generate the graph returned by visualize_tree.

Vagabond commented on src/luwak_wm_file.erl:
```diff
@@ -543,6 +545,24 @@ extract_user_meta(RD) ->
+        undefined ->
+            extract_headers(RD, T, Acc);
+        Val ->
+            extract_headers(RD, T, [{Key, Cast(Val)}|Acc])
```
Maybe do the cast in a try/catch so you can gracefully return errors if the block size isn't an integer. Right now it blows up pretty dramatically.

Vagabond commented on src/luwak_wm_file.erl:
```diff
@@ -495,7 +496,8 @@ accept_doc_body(RD, Ctx=#ctx{key=K, client=C}) ->
              _ ->
-                 {ok, H} = luwak_file:create(C, K, dict:new()),
+                 FileProps = extract_file_props(RD),
+                 {ok, H} = luwak_file:create(C, K, FileProps, dict:new()),
```
This wasn't a concern before, but now that you can pass in block-size, luwak_file:create will throw() if the block size is less than 1. You should probably catch that and convert it to a decent error message.

rzezeski committed:
Add better header validation

rzezeski commented:
I moved the header parsing to `malformed_request` where it belongs and added validation.  I think the `malformed_request` bodies could be cleaned up by using `error_resp` but that's out of scope for this pull request.

Vagabond commented:
Merged.
